### ISSUE TITLE
Managing department object view and functionality (FE)

### ISSUE THREAD

**Issue by MariusJulius** [keeleinstituut/tv-tolkevarav#180]
[Wiki](https://github.com/keeleinstituut/tv-tolkevarav/wiki/6.-Asutuse-peakasutaja-vaade#61-flow-adding--editing--deleting-institutions-%C3%BCksused-departments)
[Design - tags edit button](https://www.figma.com/file/VqCwED3Dymn7L1gGoAOK27/EKI-t%C3%B5lkev%C3%A4rav?node-id=809%3A65193&t=hCt7cWQgkDQFjnrf-1)
[Design - modal](https://www.figma.com/file/VqCwED3Dymn7L1gGoAOK27/EKI-t%C3%B5lkev%C3%A4rav?node-id=936%3A92408&t=hCt7cWQgkDQFjnrf-1) same as in #125 

BE: #179 

To-do:
- department object view and edit views and functionality
- creating views for viewing, creating, updating and deleting departments
- check privileges:
   -To add departments user needs a privilege "Asutuse üksuste lisamine" (add_department)
   -To edit departments user needs a privilege "Asutuse üksuste muutmine" (edit_department)
   -To delete departments user needs a privilege "Asutuse üksuste kustutamine" (delete_department)

**Comment by KaarelKa** [comment]
Blocked for now, has some shared parts with Tags management view + we can reuse the confirmation modal from Archiving user task. This task should be done after those.

**Comment by DanBondarenko** [comment]
Note to all developers who are working with WORKING TIMES features: see https://github.com/keeleinstituut/tv-tolkevarav/issues/212#issuecomment-1633984883.

**Comment by KaarelKa** [comment]
@VaabelMaret

1. Discuss with Inga as well, this task and the tags tasks have some shared components. Make sure you don't both work on the same things
2. We have the UserAndRoleManagementModal which should work well as the confirmation modal needed in task as well. If it does, then we should reuse it and potentially rename it to something more generic. "ConfirmationModal" or sth similar

**Comment by VaabelMaret** [comment]
@DanBondarenko  
Do I understand correctly that I can only create one new department at a time?

**Comment by KaarelKa** [comment]
@DanBondarenko Just to continue the discussion about bulk edit/delete/create for departments. Basically if it takes a lot of time on BE, then we can skip these for now, however the UX implications can be quite big. Not sure what a realistic number of potential departments can be for an institution, but playing around with some bigger numbers like 20, the delay could be quite massive for the user.

Let's say the user wants to add 10 departments, change 5 departments and delete 5 departments. This would mean 20 requests to BE after the users submits 1 form. 

Currently it seems that the response time from API is between 120ms - 200ms per request (Didn't check these requests specifically, rather just some random requests we are doing). This would mean that doing these 20 requests would take around 2.4 - 4 seconds. Actually delay would be bigger because of everything happening on the FE side related to these requests + the delays and sending of request and downloading of content. So with 20 departments, it could be a 5 second wait for the user, which is very slow according to todays standards. 



I'm not sure if we have any requirement for API response times for this project or not. @PlaksoBirgit and @MariusJulius can let us know. Often there are some requirements for max wait time for API endpoints. While this would not "technically" break any such requirements, then the experience for the end-user would still be "submitting a form takes 5 seconds" and it would break the reasoning behind such requirements

**Comment by KaarelKa** [comment]
"Why would it take 20 requests? If I understand correctly, it would be 10 requests – one for each change." -

10 X create = 10 requesta
5 X update = 5 requests
5 X delete = 5 requests

10 + 5 + 5 = 20 requests

"This assumes the requests are sent synchronously on after anothes. There is no reason to do that – the requests can be sent concurrently." - Okay true yeah, on chrome we can run 6 requests in parallel, so in case of 20 requests needed we could do it more or less in the time of 4 requests, which means the entire process would take aroung 0.8 seconds (Possibly up to 1 second, if we account for everything that needs to happen on FE). And this is with quick internet in ideal conditions (Some API responses take more than 300ms as well from time to time, so this could be a lot longer)

I think we can live with the 1 second delay. Though with 20 it would be 1 second in basically ideal circumstances (Quick internet connection + nothing affecting the speed of API responses).

Not sure how many departments could possibly exist for an institution. I would say that with 20 requests it would be on the limit of what would be an acceptable wait time (1-2 seconds with quick internet, depending on what is affecting BE API response times at the time). with slower internet (Like Fast 3G internet), the responses already take 500+ ms, so the wait time would be 2+ seconds.

And if the user decides to modify (create/update/delete) more than 20 departments at a time, then I would say the wait time would not be decent even in ideal circumstances, not to mention with slightly slower internet or possibly high load (or anything else) affecting the API


**TLDR**

I guess we can do it with the single requests for now, if it takes a lot of time on BE to add these bulk endpoints. However bigger picture it seems to me like more work on FE + worse experience for end users. So if the amount of work on BE would be the same as the increase in FE work, we are just picking the option with worse performance for no apparent reason.

**Comment by DanBondarenko** [comment]
@KaarelKa Thank you for the overview. In short: with this information, I agree that a bulk-style endpoint would be preferrable to single endpoint, but I think it’s in the interest of the project to use existing endpoints for the first phase considering our time-sensitivity.  

I don’t know what the development time cost is for writing the logic of requests sent in parallel. I suspect that it isn’t as large as the time cost for creating endpoints for bulk-modifications, especially considering the time for writing back-end tests – I would estimate this to be between 3 to 6 hours.

With regards to the 429 response – this can be configured in BE code. With that out of the way, I see this as an issue of performance optimization. I suggest creating a task to do it sometime in the future, perhaps after the project deadline.

**Comment by KaarelKa** [comment]
Yeah parallel requests are not an issue on FE, we use react-query, which easily supports this out of the box. FE time spend is rather a helper function for making sure we wait for all requests to finish and keep some state of which ones succeeded and which ones failed, so we can show the correct error messages in correct places + avoid updating the values that succeesed again, when the user resubmits fixes to errors. It's not too much of a hassle, just need to write a separate handler for this (compared to what we use for all other cases)

We just need to make sure that we configure 429 on BE side then. It should be turned off (or we should just use some really high limit) for endpoints which are used this way

**Comment by KaarelKa** [comment]
@DanBondarenko 

But yeah let's go with the parallel requests option for now. Thanks for the discussion, got some extra research done and found out some new stuff :D

**Comment by DanBondarenko** [comment]
@KaarelKa Yes, same for me – learned some important facts. Thanks! 👍

**Comment by VaabelMaret** [comment]
@DanBondarenko @KaarelKa  @MariusJulius 
I've been working on this solution for over 10 hours and it's still not done. On the FE end I can send parallel requests, I have found a way to collect the errors and show them under the inputs. But it is not easy to show successes and errors at the same time and leave the form open for the user to change the data. The form data is not updated in real time and the second request is with the old payload. This means that the user will get new errors. And if the user wants to undo the modal changes, I've to send another payload to restore the old state. 
I think it will take another 5-10 hours to find the best cancel solution and remove the errors from the already updated values.
I think it's much faster if BE does the same endpoint to add/delete/update the department as we have for updating tags.

**Comment by VaabelMaret** [comment]
@DanBondarenko @KaarelKa 
I understand it's busy times for BE at the moment, and I've come up with a workaround for FE that works and notifies the client. However I find this is not the best solution for how edit and undo edits should work.

**Comment by KaarelKa** [comment]
@MariusJulius Hey can you create a followup task to this one. The followup task would be for changing how departments are changed/deleted/changed on both FE and BE

**Current solution**:

1. FE side maps all the changes made in the modal
2. A request is made for each changed department (List of PUT/POST/DELETE requests in a row for create/edit/delete actions)
3. FE wait for all of these requests to finish
4. FE maps any errors we got from the requests and shows the errors in the Modal under input fields
5. All the requests that passed are done
6. User makes changes to errored fields and can submit again

This kind of works for now, but currently the "Tühista" button does not work correctly. When the user presses "Tühista" after the first submit, they will not end up with the expected result.

Expected: All the departments are reset to the state they had before opening the modal

Actual: Only the departments that got some BE error will be reset and the other changes have gone through

With the current approach FE would need to map a list of reverse queries to undo all of the previously done changes. This would take quite a bit of time and has 1 extra issue. To revert deletion of a department, we would need to send the id of that department to BE + we would need to add extra logic to BE, so the department could be attached to all the correct places. This would take more time on both FE and BE than just using the same logic for adding/changing/deleting departments as is used for adding/changing/deleting tags.

**New tasks/New solution**

1. BE - Add same type of bulk endpoints for departments as are used for tags
2. FE - Use those endpoints the same way they are used for tags and remove current custom workarounds.

**Comment by MariusJulius** [comment]
Tasks:
#439 
#440

**Comment by plakitkelly** [comment]
Tested in 27.09. It's OK
